[3.x] AnimatedSprite Fix updating inspector when SpriteFrames is modified

kleonc opened the pull request:
(needs a separate `master` version due to #45879; for the same reason some issues fixed here might already be non-existent in the `master`; haven't checked none of these)

Makes creating / renaming / deleting animations in the `SpriteFramesEditorPlugin` properly update properties of the `AnimatedSprite` in the inspector (mainly options in the animation dropdown). Removing current animation of the `AnimatedSprite` will make the first (after removal) animation left be selected as the new current animation (if no animation would be left then nothing will be done from the inspector point of view). 


Reverts #27254 but keeps #26381 being fixed.
Fixes #44438.
Fixes #46741.
Fixes #28658, fixes #40263 (these look like duplicates).

Removes unused `SpriteFrames::_get_animation_list()` method.

~~After reverting #27254 newly created animation is no longer automatically set as the current animation of the `AnimatedSprite`/`AnimatedSprite3D` being edited. Should the current behavior be preserved (assigning it automatically)?~~
I've changed it so currently the newly created animation is automatically set as the current animation of the anim sprite being edited (same as before), and the same for all other anim sprites having currently invalid animation set (same as before). But valid animations of other anim sprites will no longer be automatically changed (as they used to be).

kleonc committed:
AnimatedSprite Fix updating inspector when SpriteFrames is modified

KoBeWi commented:
I checked the issues. One is valid on both branches and one only in 3.x

> newly created animation is no longer automatically set as the current animation of the AnimatedSprite/AnimatedSprite3D being edited.

I think the reason for this was that when you create new animation and put some sprites, you can preview it immediately.

kleonc commented:
> I checked the issues. One is valid on both branches and one only in 3.x

Hmm, I think I've listed 3 issues (4 if counting the duplicated one): 

> Reverts #27254 but keeps #26381 being fixed.
> Fixes #44438.
> Fixes #46741.
> Fixes #28658, fixes #40263 (these look like duplicates).

so it's kinda hard to tell what have you found out. :smile: So please try being more specific. Also thanks for checking it out!
Oh, and if by any chance you'd like to do `master` version feel free to do so (or anyone else).
 
---


> > newly created animation is no longer automatically set as the current animation of the AnimatedSprite/AnimatedSprite3D being edited.
> 
> I think the reason for this was that when you create new animation and put some sprites, you can preview it immediately.


For me it looks like it was just an unintended side effect of #27254. But yeah, I can see how it might be a desired behavior for some people and that's why I've asked which behavior is the desired one / should current behavior be preserved.

KoBeWi commented:
> so it's kinda hard to tell what have you found out. 

I meant #28658 and #46741 (the other ones are closed now).

akien-mga commented on scene/2d/animated_sprite.cpp:
```diff
@@ -144,20 +144,24 @@ void SpriteFrames::clear(const StringName &p_anim) {
 void SpriteFrames::clear_all() {
 	animations.clear();
 	add_animation("default");
+	// emit_changed(); will be triggered by the previous line, no need to do it twice
```
Makes it look less like dead code to remove.
```suggestion
	add_animation("default"); // Also emits changed.
```

kleonc commented:
@akien-mga Rebased, applied suggestion. Added some more comments and changed this (part of the PR description):

> ~After reverting #27254 newly created animation is no longer automatically set as the current animation of the `AnimatedSprite`/`AnimatedSprite3D` being edited. Should the current behavior be preserved (assigning it automatically)?~
> I've changed it so currently the newly created animation is automatically set as the current animation of the anim sprite being edited (same as before), and the same for all other anim sprites having currently invalid animation set (same as before). But valid animations of other anim sprites will no longer be automatically changed (as they used to be).

akien-mga reviewed (approved):
Should be good to merge after 3.5 is released.

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 3.5.1.
